examples/vnet/doca_gw_dpdk.c

```diff
@@ -0,0 +1,265 @@
+		if (is_ipv6_max(src_ip.a.ipv6_addr))
+			param->flags |= (type == OUTER_MATCH ? DOCA_OUTER_SRC_IP : DOCA_INNER_SRC_IP);
+	}
+
+	if (!is_ipv6_zero(dst_ip.a.ipv6_addr)){
+		memcpy(spec->hdr.dst_addr, dst_ip.a.ipv6_addr, sizeof(dst_ip.a.ipv6_addr));
+		if (is_ipv6_max(dst_ip.a.ipv6_addr))
+			param->flags |= (type == OUTER_MATCH ? DOCA_OUTER_DST_IP : DOCA_INNER_DST_IP);
+	}
+
+	if (param->flags) {
+		item->spec = spec;
+		item->mask = mask;
+		param->apply = doca_gw_dpdk_set_ipv6_flow_item;
+	}
+}
+
+static int doca_gw_dpdk_set_vxlan_flow_item(__rte_unused struct gw_item_param *param,
+						__rte_unused struct doca_gw_match *match)
+{
+	return 0;
+}
+
+static void doca_gw_dpdk_build_vxlan_flow_item(struct gw_item_param *param, struct doca_gw_match *match)
+{
+	struct rte_flow_item *item = param->item;
+	struct rte_flow_item_vxlan *spec = &param->item_param.vxlan.spec;
+	struct rte_flow_item_vxlan *mask = &param->item_param.vxlan.mask;
+
+	item->type = RTE_FLOW_ITEM_TYPE_VXLAN;
+	if (match->tun.vxlan.tun_id) {
+		//memcpy(spec->vni, match->tun.vxlan.tun_id, 3);
+		if (match->tun.vxlan.tun_id == UINT32_MAX)
+			param->flags |= DOCA_VXLAN_VNI;
+	}
+
+	if (param->flags) {
+		item->spec = spec;
+		item->mask = mask;
+		param->apply = doca_gw_dpdk_set_vxlan_flow_item;
+	}
+}
+
+static void doca_gw_dpdk_build_tcp_flow_item(struct gw_item_param *param,
+			struct doca_gw_match *match, uint8_t type)
+{
+	struct rte_flow_item *item = param->item;
+	struct rte_flow_item_tcp *spec = &param->item_param.tcp.spec;
+	struct rte_flow_item_tcp *mask = &param->item_param.tcp.mask;
+
+	item->type = RTE_FLOW_ITEM_TYPE_TCP;
+	spec->hdr.dst_port = (type == INNER_MATCH)? match->in_src_port : match->out_src_port;
+	spec->hdr.dst_port = (type == INNER_MATCH)? match->in_src_port : match->out_src_port;
+	item->spec = spec;
+	item->mask = mask;
+	//no need modify those fields.
+	return ;
+}
+
+static void doca_gw_dpdk_build_udp_flow_item(struct gw_item_param *param,
+			struct doca_gw_match *match, uint8_t type)
+{
+	struct rte_flow_item *item = param->item;
+	struct rte_flow_item_udp *spec = &param->item_param.udp.spec;
+	struct rte_flow_item_udp *mask = &param->item_param.udp.mask;
+
+	item->type = RTE_FLOW_ITEM_TYPE_UDP;
+	spec->hdr.dst_port = (type == INNER_MATCH)? match->in_src_port : match->out_src_port;
+	spec->hdr.dst_port = (type == INNER_MATCH)? match->in_src_port : match->out_src_port;
+	item->spec = spec;
+	item->mask = mask;
+	//no need modify those fields.	
+	return ;
+}
+
+static int doca_gw_dpdk_build_item(struct doca_gw_match *match,
+					struct gw_pipe_dpdk_flow *pipe_flow)
+{
+	uint8_t idx = 0, type = OUTER_MATCH;
+
+	doca_gw_dpdk_build_eth_flow_item(&pipe_flow->item_param[idx++], match);
+	if (match_is_ipv4(match, OUTER_MATCH))
+		doca_gw_dpdk_build_ipv4_flow_item(&pipe_flow->item_param[idx++], match, type);
+	else
+		doca_gw_dpdk_build_ipv6_flow_item(&pipe_flow->item_param[idx++], match, type);
```

Comment:
you need to test also (out_l3_type is a typo, should be l4_type)
    uint8_t  out_l3_type;
    uint16_t out_src_port;
    uint16_t out_dst_port;

because user might have overlay to underlay, so he will match on 5-tuple with no tunnel

Reply:
yes,  if no tunnel , " type = OUTER_MATCH" also, will  build the l4 items.